src/v/kafka/client/test/fixture.h

```diff
@@ -74,4 +74,51 @@ class kafka_client_fixture : public redpanda_thread_fixture {
         add_topic(tp_ns, partitions).get();
         return tp_ns;
     }
+
+    void enable_sasl_and_restart(ss::sstring username) {
+        ss::smp::invoke_on_all([username]() mutable {
+            auto& config = config::shard_local_cfg();
+
+            config.get("enable_sasl").set_value(true);
+            config.get("superusers")
+              .set_value(std::vector<ss::sstring>{username});
+
+            auto& node_config = config::node();
+            int32_t kafka_port
+              = node_config.kafka_api.value()[0].address.port();
+            node_config.get("kafka_api")
+              .set_value(std::vector<config::broker_authn_endpoint>{
+                config::broker_authn_endpoint{
+                  .address = net::unresolved_address("127.0.0.1", kafka_port),
+                  .authn_method = config::broker_authn_method::sasl}});
+
+            node_config.get("admin").set_value(
+              std::vector<model::broker_endpoint>{model::broker_endpoint(
+                net::unresolved_address("127.0.0.1", 9644))});
+        }).get();
+
+        restart();
+    }
+
+    void disable_sasl_and_restart() {
+        ss::smp::invoke_on_all([]() mutable {
+            auto& config = config::shard_local_cfg();
+
+            config.get("enable_sasl").reset();
+            config.get("superusers").reset();
+
+            auto& node_config = config::node();
+            int32_t kafka_port
+              = node_config.kafka_api.value()[0].address.port();
+            node_config.get("kafka_api")
+              .set_value(std::vector<config::broker_authn_endpoint>{
+                config::broker_authn_endpoint{
+                  .address = net::unresolved_address("127.0.0.1", kafka_port),
+                  .authn_method = std::nullopt}});
+            node_config.get("admin").set_value(
+              std::vector<model::broker_endpoint>{});
+        }).get0();
+
+        restart();
```

Comment:
After #6876 merged, this unit test started failing with `heap-buffer-overflow` at this call to `restart()`.
Here is the full stack trace from my local (I rebased dev to get the updated code):
```
    #0 0x55b923eb3927 in cluster::controller::get_feature_table() /home/nyalia/redpanda/src/v/cluster/controller.h:105:16
    #1 0x55b93170d7d5 in features::feature_migrator::start(seastar::abort_source&) /home/nyalia/redpanda/src/v/features/feature_migrator.cc:38:47
    #2 0x55b924d263e6 in application::start_runtime_services(cluster::cluster_discovery&) /home/nyalia/redpanda/src/v/redpanda/application.cc:1651:12
    #3 0x55b924d1ed9f in application::wire_up_and_start(stop_signal&, bool) /home/nyalia/redpanda/src/v/redpanda/application.cc:1454:5
    #4 0x55b9243c0a9c in kafka_client_fixture::restart() /home/nyalia/redpanda/src/v/kafka/client/test/fixture.h:31:13
    #5 0x55b9243c587b in kafka_client_fixture::disable_sasl_and_restart() /home/nyalia/redpanda/src/v/kafka/client/test/fixture.h:122:9
```

We see there is a call to `migrator.start()` which fetches the feature table but apparently that memory address does-not-exist yet. Based on the follow code block in wire_up_and_start it seems to me that the migrators vector is expected to be empty:
```
    if (test_mode) {
        // When running inside a unit test fixture, we may fast-forward
        // some of initialization that would usually wait for the controller
        // to commit some state to its log.
        vlog(_log.warn, "Running in unit test mode");
        _feature_table
          .invoke_on_all(
            [](features::feature_table& ft) { ft.testing_activate_all(); })
          .get();
    } else {
        // Only populate migrators in non-unit-test mode
        _migrators.push_back(
          std::make_unique<features::migrators::cloud_storage_config>(
            *controller));
        _migrators.push_back(
          std::make_unique<features::migrators::group_metadata_migration>(
            *controller, group_router));
    }
```

@jcsp I think you made changes to this code recently, do you have any recommendation to get over this? I thought about clearing the migrators vector or disabling the offending line when `test_mode=true` but is that OK? Do all unit tests need the migrators vector?

Reply:
To be more clear, the offending block of code is within `start_runtime_services`:
```cpp
    for (const auto& m : _migrators) {
        m->start(controller->get_abort_source().local());
    }
```
The migrators vector was **not empty** here even though I think it was supposed to be empty.

Reply:
> Do all unit tests need the migrators vector?

It's the opposite: unit tests have the special case to just switch all the features on, non-unit-test mode is where migrators get instantiated because we might be running out in the real world on an upgrading system.

Reply:
It's actually not obvious to me why this crashes -- I don't see any special casing in the unit test that should cause us to reach the migrator startup without the feature table + controller having been constructed.  But given that migrators are never needed in a fixture test, I'd rather just have the fixture tests switch them off via `test_mode` than figure it out.